.cascade-code/Chapman.edu/test-section/tatwell/query_api_spike/query_api.vtl

```diff
@@ -0,0 +1,175 @@
+ * query_api.vtl
+ * Test format to demonstrate usage of Cascade locator tool and Query API.
+ *
+ * For more info, see spike user story:
+ * https://trello.com/c/InwKZ4w9
+ *#
+
+#import( "/_cascade/formats/helpers/debug.velocity" )
+
+##
+## Hardcoded Params
+##
+
+
+##
+## Vars
+##
+## Cascade provides a $currentPage object to every format.
+
+## This can be selected when Page Block is present:
+##set ( $pageNode = $_XPathTool.selectSingleNode($contentRoot, "//system-page[@current = 'true']") )
+
+
+##
+## Macros
+##
+#macro ( debugCurrentPage )
+  <pre>
+  # currentPage
+  currentPage name: $currentPage.name
+  currentPage path: $currentPage.path
+  currentPage assetType: $currentPage.assetType
+  currentPage Hide from navigation?: $currentPage.metadata.getDynamicField('Hide from navigation').checkbox
+
+  # _PropertyTool.outputProperties( currentPage )
+  $_PropertyTool.outputProperties( $currentPage )
+  </pre>
+#end
+
+#macro ( debugPathTools )
+  #set ( $currentPath = $currentPage.path )
+  #set ( $pathList = $currentPath.split('/') )
+  #set ( $parentIndex = $pathList.size() - 2 )
+  #set ( $parentFolderName = $pathList.get($parentIndex) )
+  #set ( $grandParentPath = $_StringTool.substringBefore($currentPath, $parentFolderName) )
+
+  <pre>
+  # debugPathTools
+  currentPage.parentFolder: $currentPage.parentFolder
+  \- isNull: $_PropertyTool.isNull($currentPage.parentFolder)
+  currentPath: $currentPath
+  pathList.size: $pathList.size()
+  parentFolderName: $parentFolderName
+  grandParentPath: $grandParentPath
+  </pre>
+#end
+
+#macro ( queryCurrentPageParentSiblings )
+  ## Note: Velocity (Java) arrays don't have a pop or remove method.
+  #set ( $parentSiblings = [] )
+
+  #set ( $currentPath = $currentPage.path )
+  #set ( $pathList = $currentPath.split('/') )
+  #set ( $parentIndex = $pathList.size() - 2 )
+  #set ( $parentFolderName = $pathList.get($parentIndex) )
+
+  ## locateFolder fails with trailing slashes. Thus the leading slash added here.
+  ## This will also need some sanity checks for paths less than 3 folders deep.
+  #set ( $grandParentPath = $_StringTool.substringBefore($currentPath, "/${parentFolderName}") )
+  #set ( $grandParentFolder = $_.locateFolder($grandParentPath) )
+
+  #foreach( $parentSibling in $grandParentFolder.children )
+    ## add method returns boolean which will displayed if not assigned to var.
+    #set ( $discard = $parentSiblings.add($parentSibling) )
+  #end
+
+  <pre>
+  # queryCurrentPageParentSiblings
+  currentPath: $currentPath
+  grandParentPath: $grandParentPath
+  grandParentFolder isNull?: $_PropertyTool.isNull($grandParentFolder)
+  parentSiblings: ${_DisplayTool.list($parentSiblings,"/")}
+  grandParentFolder.path: $grandParentFolder.path
+
+  # ParentSiblings: title : visible?
+  #foreach( $parentSibling in $grandParentFolder.children )
+    #set ( $metaHide = $parentSibling.metadata.getDynamicField('Hide from navigation').value )
+    #set ( $isVisible = $metaHide != 'Yes' )
```

Comment:
I'd like to believe there is a cleaner way to do this. But this seems to work.

Reply:
I noticed this with some of the folders when I was testing: the folder will be visible but the index hidden.

Is there any reason we can't require the folders to be marked hidden in Cascade?

I think that convention would be simpler and more straightforward than having the code do multiple checks.

Reply:
We do tell people to mark the folder as hidden but they may or may not always do that, opting to mark the index page instead (or as well). The current code was written to catch both cases. Marking the index page would be the same (but bad practice) as marking the folder, indicating the whole folder should still be hidden (if you can't see the landing page you shouldn't see the children). We could see if we could generate a report of folders and pages marked as hidden and do a big cleanup so it's all done correctly, and then when your program is modified it could work as you suggest. Anyone doing it wrong going forward would see that the intended landing page didn't get hidden as they thought and fix quickly (I would think). We should run this by Ross to make sure he's ok with that change in behavior (which really was the original intent but people are people and don't always do what we'd prefer

Reply:
I like this idea. With the Off Canvas Nav, it should be simpler as it only looks at Umbrella Category folders.

With the Left Nav, yes, I'd prefer to check only folders. But since this involves more end-user responsibility, I'd be more flexible with including the index file check there. We can look it at more closely when we get to that story.